Enable preview=dtorFields by default

MoonlightSentinel opened the pull request:
~~DMD test suite passed locally, so~~ let's see what breaks.

CC @wilzbach, @Geod24

MoonlightSentinel committed:
Enable preview=dtorFields by default

The preview was introduced in 2.075 and should be enabled by default.

wilzbach commented on src/dmd/cli.d:
```diff
@@ -756,7 +756,7 @@ dmd -cov -unittest myprog.d
             "fix integral promotions for unary + - ~ operators"),
         Feature("dtorfields", "dtorFields",
-            "destruct fields of partially constructed objects"),
+            "destruct fields of partially constructed objects", false, false),
```
You'll have to update `test/compilable/previewhelp.d` as it's now no longer documented.

MoonlightSentinel replied:
Definitly should have remembered that...

MoonlightSentinel commented:
Well, this will cause some breakage in the wild because of a possible attribute missmatch for ctor & dtor (as seen in `astbase.d`). But some of that could be averted if we enable attribute interference for generated destructors (e.g. `RootObject.~this` is currently `@system`).

MoonlightSentinel commented:
Actually, inference should work for generated dtors. But there seems to be something funky going on:

```
class Parent
{
    this() nothrow pure @nogc @safe {}
}

class Child : Parent
{
    S s;
    // this() {} // Uncomment this to make the test pass
}

struct S
{
    ~this() {}
}

```

MoonlightSentinel commented:
Found the problem. Generated subclass constructors simply copied the attributes from the superclass constructor. This worked fine as long as the body only contained `super(...)` but obviously fails when adding a dtor call.

MoonlightSentinel commented:
Does it make sense to introduce a transition-phase, during which the additional destructor call may violate the constructors attributes, and issue a deprecation message whenever a violation is suppressed?

MoonlightSentinel commented:
`eventcore` fails because the constructors of several classes are `@nogc` but their `destructors` aren't...

wilzbach commented:
> Does it make sense to introduce a transition-phase, during which the additional destructor call may violate the constructors attributes, and issue a deprecation message whenever a violation is suppressed?
> 
> EDIT: When not explicitly passing `-preview=dtorfields`

Yes, looking at the Buildkite projects this most definitely would help a lot to prevent breakage in the wild.

Geod24 commented:
> Does it make sense to introduce a transition-phase

What exactly do you have in mind ? From what I can see, something that would say:
```
Deprecation: `$CLASS_FQN.this(int, string) @safe @nogc` attributes are stricter than its destructors attributes (`@safe`)
             Make constructor `nothrow`, adjust destructors attributes, or remove the destructor
```

Is that it ?

MoonlightSentinel commented:
> Yes, looking at the Buildkite projects this most definitely would help a lot to prevent breakage in the wild.

Agreed, but BuildKite is somewhat misleading as most projects are failing due to their dependency on `eventcore` and `dsymbol` AFAICT.

> ```
> Deprecation: `$CLASS_FQN.this(int, string) @safe @nogc` attributes are stricter than its destructors attributes (`@safe`)
>              Make constructor `nothrow`, adjust destructors attributes, or remove the destructor
> ```
> Is that it ?

Yes, except that it should also mention why this change is required ("might call destructor if an exception is thrown")

MoonlightSentinel commented:
Including #11692 reduces the breakage on BuildKite to destructor calls for nested `RefCounted` members inserted into `@safe` constructors ...

wilzbach commented:
> Including #11692 reduces the breakage on BuildKite to destructor calls for nested `RefCounted` members inserted into `@safe` constructors ...

Sorry. It's merged now, so you likely need to rebase this due to conflicts.

MoonlightSentinel commented:
No worries, just merged it into this PR to estimate the effect on BuildKite.

MoonlightSentinel committed:
Allow pure, @safe, @nogc, violations for inserted dtor call...

... but issue a deprecation instead. This should allow for a smoother
transition without creating too much friction.

Note that this excludes `nothrow` because the dtor call is omitted for
`nothrow` functions.

MoonlightSentinel commented:
The second commit suppresses errors for missmatched attributes ( `@nogc´, `pure`, `@safe`) and issues a deprecation instead.

wilzbach commented on test/compilable/dtorfields_deprecation.d:
```diff
@@ -0,0 +1,49 @@
+/**
+Checks that code still compiles when -preview=default is enabled by default
```
preview=dtorfields

wilzbach commented on test/compilable/riia_ctor.d:
```diff
@@ -11,23 +11,6 @@ class C
-{
-    Array arr;
-    this() @safe {}
-}
```
Move into fail_compilation?

MoonlightSentinel replied:
Reverted this and (ab)used it to test `-revert=dtorfields` ... which fails ATM.

MoonlightSentinel replied:
`parseCLIOption` doesn't handle reverts correctly. This is the generated code:

```
case "all":
        params.noDIP25 = true;
        params.markdown = true;
        params.dtorFields = true;
        return true;
case "dip25":
        params.noDIP25 = true;
        return true;

case "markdown":
        params.markdown = true;
        return true;

case "dtorfields":
        params.dtorFields = true;
        return true;
```

`-revert=dip25` only works because of the second option `noDIP25`. I think a better approach would be to  store  revertable switches as a three state boolean (e.g `set`, `default`, `unset`)

MoonlightSentinel commented:
Should this distinguish between whether the user has activated the flag or not?

- Raise a deprecation if it was enabled by default
- Error if the user passed `-preview=dtorfields` explicitly

Then  users can opt-in to early without needing `-de` and the behaviour for projects already using that preview remains unchanged.

wilzbach commented:
I would keep it simple and issue a deprecation in both cases (same behavior). Phobos and druntime are already using `-de`.

MoonlightSentinel commented:
Not sure if `-de` is feasible for all projects. Also there is a precedent with `-preview=dip25` showing the same behaviour.

MoonlightSentinel commented:
Added a changelog entry.

RazvanN7 commented:
cc @wilzbach @Geod24

nordlow commented:
- Nitpick: dtorFields => dtorfields in title
- Are there any useful patterns that are enabled/made safe with this change?

PetarKirov commented:
@nordlow Are there any useful patterns that are enabled/made safe with this change?

Although this may seem like a niche change, without it D can't claim to fully implement RAII, as it's at the core of having correct transactional semantics for constructors.

nordlow commented on changelog/dtorfileds.md:
```diff
@@ -0,0 +1,33 @@
+when an exception is thrown inside of an constructor. It was introduced in
+2.075 and is now enabled by default.
+
+Note that code may fail to compile if a constructor with strict attributes
```
Replace with:

"...constructor, such as `Scanner.this()` below ..."

Geod24 commented:
This PR introduced a regression: https://issues.dlang.org/show_bug.cgi?id=21989
